### PR TITLE
Reworked maybe_grow to take a single ptr

### PR DESCRIPTION
**Pull request by domenukk**
This PR changes maybe_grow to take a single pointer and adds a function to free it also.
I'd be glad to hear y'alls opinions.

### CODE REVIEW

File: afl-cmin [hunk 1]
```diff
@@ -1,6 +1,5 @@
 #!/usr/bin/env sh
 export AFL_QUIET=1
-export ASAN_OPTIONS=detect_leaks=0
```
Comment (vanhauser-thc) [inline afl-cmin]: why did you remove this? otherwise targets compiled with asan produce errors.
that an asan compiled showmap does not have leaks we can test with alf-showmap directly

Comment (domenukk) [reply afl-cmin]: How do you mean? We don't care about targets here as this is the aflpp makefile?

File: examples/afl_network_proxy/afl-network-server.c [hunk 2]
```diff
@@ -379,14 +380,15 @@ int recv_testcase(int s, void **buf, size_t *max_len) {
...
+    buf2 = maybe_grow((void **)&buf2, clen);
+    if (unlikely(!buf2)) { PFATAL("Alloc"); }
     received = 0;
     while (received < clen &&
            (ret = recv(s, buf2 + received, clen - received, 0)) > 0)
       received += ret;
     if (received != clen) FATAL("did not receive compressed information");
     if (libdeflate_deflate_decompress(decompressor, buf2, clen, (char *)*buf,
-                                      *max_len,
+                                      maybe_grow_bufsize(buf2),
```
Comment (vanhauser-thc) [inline examples/afl_network_proxy/afl-network-server.c]: this is a different parameter. either this is now wrong, or I had a bug in it. too tired atm to analyse this :(

Comment (domenukk) [reply examples/afl_network_proxy/afl-network-server.c]: Yes I wondered why you were using the internal size here.

File: include/alloc-inl.h [hunk 3]
```diff
@@ -673,78 +619,96 @@ static inline size_t next_pow2(size_t in) {
...
 
-/* This function makes sure *size is > size_needed after call.
- It will realloc *buf otherwise.
- *size will grow exponentially as per:
- https://blog.mozilla.org/nnethercote/2014/11/04/please-grow-your-buffers-exponentially/
- Will return NULL and free *buf if size_needed is <1 or realloc failed.
- @return For convenience, this function returns *buf.
- */
-static inline void *maybe_grow(void **buf, size_t *size, size_t size_needed) {
+struct maybe_grow_buf {
```
Comment (vanhauser-thc) [inline include/alloc-inl.h]: this is now a function with just two variables and nothing else?

File: include/alloc-inl.h [hunk 4]
```diff
@@ -673,78 +619,96 @@ static inline size_t next_pow2(size_t in) {
...
+};
 
-  /* grow exponentially */
-  size_t next_size = next_pow2(size_needed);
+#define SIZE_OFFSET offsetof(struct maybe_grow_buf, buf)
 
-  /* handle overflow and zero size_needed */
-  if (!next_size) { next_size = size_needed; }
+/* Returs the container element to this ptr */
+static inline struct maybe_grow_buf *maybe_grow_buf(void *buf) {
```
Comment (vanhauser-thc) [inline include/alloc-inl.h]: this function returns the buffer pointer right? the name sounds rather like you want to grow a buffer.
recommendation: rename to maybe_grow_get_bufptr and the one below maybe_grow_get_bufsize

File: include/alloc-inl.h (hunk 3 above)
Comment (domenukk) [reply include/alloc-inl.h]: Aah no this is just the internal struct used by the actual function to calculate offsets in a less.obscure way :)

File: include/alloc-inl.h [hunk 5]
```diff
@@ -673,78 +619,96 @@ static inline size_t next_pow2(size_t in) {
...
  It will realloc *buf otherwise.
  *size will grow exponentially as per:
  https://blog.mozilla.org/nnethercote/2014/11/04/please-grow-your-buffers-exponentially/
- Will FATAL if size_needed is <1.
+ Will return NULL and free *buf if size_needed is <1 or realloc failed.
  @return For convenience, this function returns *buf.
  */
-static inline void *ck_maybe_grow(void **buf, size_t *size,
-                                  size_t size_needed) {
+static inline void *maybe_grow(void **buf, size_t size_needed) {
```
Comment (vanhauser-thc) [inline include/alloc-inl.h]: that this is the function that actually increases a buffer is not clear in contrast to the other names.
maybe rename this to maybe_grow_need or _enlarge, or ...

File: include/alloc-inl.h [hunk 6]
```diff
@@ -673,78 +619,96 @@ static inline size_t next_pow2(size_t in) {
...
 
   /* No need to realloc */
-  if (likely(*size >= size_needed)) { return *buf; }
+  if (likely(size >= size_needed)) { return *buf; }
 
   /* No initial size was set */
   if (size_needed < INITIAL_GROWTH_SIZE) { size_needed = INITIAL_GROWTH_SIZE; }
 
   /* grow exponentially */
   size_t next_size = next_pow2(size_needed);
```
Comment (vanhauser-thc) [inline include/alloc-inl.h]: you coult put this together:
if  (size_needed < INITIAL_GROWTH_SIZE) { size_needed = INITIAL_GROWTH_SIZE; next_size = INITIAL_GROWTH_SIZE << 1; }
else next_size = next_pow2(size_needed);
this way if <= this is precomputed

File: include/alloc-inl.h (hunk 5 above)
Comment (domenukk) [reply include/alloc-inl.h]: Thats what `grow` was supposed to say 😃
Probably we want a better naming scheme overall?

File: include/alloc-inl.h [hunk 7]
```diff
@@ -673,78 +619,96 @@ static inline size_t next_pow2(size_t in) {
...
 
-  /* handle overflow */
+  /* handle overflow and zero size_needed */
   if (!next_size) { next_size = size_needed; }
```
Comment (vanhauser-thc) [inline include/alloc-inl.h]: would this be more secure as `if (next_size < size_needed)` ?

File: include/alloc-inl.h [hunk 8]
```diff
@@ -30,12 +30,13 @@
...
 #include <string.h>
+#include <stddef.h>
 
 #include "config.h"
 #include "types.h"
 #include "debug.h"
 
-/* Initial size used for ck_maybe_grow */
+/* Initial size used for maybe_grow */
 #define INITIAL_GROWTH_SIZE (64)
```
Comment (vanhauser-thc) [inline include/alloc-inl.h:41]: the ->size value in the structure could either be the full size of the allocated memory, or the size available to the user.
currently it is the previous.

if ->size is just internal it does not matter and whatever results in the fewest computations required should be it.

if this is given to the user it should be the latter as otherwise a user has to know do deduct a pointer size from the value and forgetting that would lead to overruns.

File: src/afl-fuzz-one.c [hunk 9]
```diff
@@ -364,7 +364,7 @@ static void locate_diffs(u8 *ptr1, u8 *ptr2, u32 len, s32 *first, s32 *last) {
 
 #endif                                                     /* !IGNORE_FINDS */
 
-#define BUF_PARAMS(name) (void **)&afl->name##_buf, &afl->name##_size
+#define BUF_PARAMS(name) (void **)&afl->name##_buf
```
Comment (vanhauser-thc) [inline src/afl-fuzz-one.c]: if we would call this AFL_BUF_PARAM we could put it in types.h and never would need to define + undefine it.

Comment (vanhauser-thc) [reply src/afl-fuzz-one.c]: and the S at the end can be removed anyway as it not only the buf param, and not also the size

File: src/afl-fuzz-one.c [hunk 10]
```diff
@@ -4339,7 +4351,8 @@ static u8 mopt_common_fuzzing(afl_state_t *afl, MOpt_globals_t MOpt_globals) {
 
               if (temp_len + extra_len >= MAX_FILE) break;
 
-              out_buf = ck_maybe_grow(BUF_PARAMS(out), temp_len + extra_len);
+              out_buf = maybe_grow(BUF_PARAMS(out), temp_len + extra_len);
+              if (unlikely(!out_buf)) { PFATAL("alloc"); }
```
Comment (vanhauser-thc) [inline src/afl-fuzz-one.c:4354]: if we move the PFATAL into the .h it cannot be forgotten and less code lines overall

**Comment by vanhauser-thc** [comment]
done :)

File: src/afl-fuzz-one.c (hunk 10 above)
Comment (domenukk) [reply src/afl-fuzz-one.c:4354]: For multithreaded fuzzing we won't be able to use PFATAL anymore, so I suggest slowly reworking the error handling

File: afl-cmin (hunk 1 above)
Comment (vanhauser-thc) [reply afl-cmin]: we dont care about leaks in targets is what I means :) and with your removal you enable printing+erroring on them.
its in the afl-cmin script

File: include/alloc-inl.h (hunk 5 above)
Comment (vanhauser-thc) [reply include/alloc-inl.h]: could also be a different naming scheme. having grow in every function name might be misleading :)

File: src/afl-fuzz-one.c (hunk 10 above)
Comment (vanhauser-thc) [reply src/afl-fuzz-one.c:4354]: ah good point, yes then not moving FATAL.
however we have a disadvantage. we only need a check if a realloc was done. however the caller doesnt know if one was done or not and therefore always has to check. but whelp I guess that is the disadvanted of threading

Comment (domenukk) [reply src/afl-fuzz-one.c:4354]: As maybe_grow is inlined the compiler can figure that out, but I can double check

File: afl-cmin (hunk 1 above)
Comment (domenukk) [reply afl-cmin]: ooh oops that was a fail :)

File: include/alloc-inl.h (hunk 7 above)
Comment (domenukk) [reply include/alloc-inl.h]: This is impossible given the current next_pow2 output. I added a unittests for that just in case.

**Comment by domenukk** [comment]
Renamed `maybe_grow` to `afl_realloc`
